### PR TITLE
StreamMapIterator (was asyncmap)

### PR DESCRIPTION
**Pull request by samoconnor**
First step in [Simplifying and generalising pmap (as per  #14843)](https://github.com/JuliaLang/julia/issues/14843).

Please limit comments on this PR to technical / implementation detail relating to the code in the PR.

Please make comments relating the interface and the overall pmap refactoring in issue #14843.

### CODE REVIEW

File: base/mapiterator.jl [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+"""
+imap(f, c...; ntasks=nothing) = StreamMapItr(f, c...; ntasks=ntasks)
+
+
+"""
+    amap!(function, collection; ntasks=100)
+
+In-place version of [`amap`](:func:`amap`).
+"""
+function amap! end
```
Comment (tkelman) [inline base/mapiterator.jl]: This should have an actual signature so the docstring gets disambiguated.

File: base/mapiterator.jl [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+"""
+    amap!(function, destination, collection...; ntasks=100)
+
+Like [`amap`](:func:`amap`), but stores the result in `destination` rather than a new collection.
+`destination` must be at least as large as the first collection.
+"""
+function amap!(f, c...; ntasks=nothing)
+
+    destination = c[1]
```
Comment (tkelman) [inline base/mapiterator.jl]: The signature shouldn't use splatting in this way, providing only one input should be a `MethodError` not a `BoundsError`

File: base/mapiterator.jl [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+type AsyncMapItr
+    f
+    results
+    arg_enum::Enumerate
+    ntasks::Int
+end
+
+function AsyncMapItr(f, results, c...; ntasks=nothing)
+    if ntasks == nothing
```
Comment (tkelman) [inline base/mapiterator.jl]: just default it to 100 in the signature?

File: base/exports.jl [hunk 4]
```diff
@@ -1210,8 +1210,11 @@ export
 
 # multiprocessing
     addprocs,
+    amap,
+    amap!,
     ClusterManager,
     fetch,
+    imap,
```
Comment (tkelman) [inline base/exports.jl]: imap needs tests if it's going to be exported

Comment (samoconnor) [inline base/exports.jl]: `amap(f, c...; kv...) = collect(imap(f, c...; kv...))`, and `amap` is subjected to the same tests as `map`, ...

File: base/mapiterator.jl (hunk 3 above)
Comment (samoconnor) [inline base/mapiterator.jl]: There are a whole stack of calls that pass `nothing` as the default.
Don't want to have to change multiple places when changing the default.

Comment (tkelman) [inline base/mapiterator.jl]: Give it a name then? Using `nothing` like this is bad for type stability.

File: base/mapiterator.jl (hunk 2 above)
Comment (samoconnor) [inline base/mapiterator.jl]: fair enough, I'll change it to something like this:

```julia
function amap!(f, destination, c...; ntasks=nothing)

    if length(c) == 0
        c = [destination]
    end

    for task in AsyncMapItr(f, destination, c..., ntasks=ntasks) end

    return destination
end
```

Comment (tkelman) [inline base/mapiterator.jl]: That's not great either. Use dispatch with separate methods for the 2-arg case vs the n-arg case, as in https://github.com/JuliaLang/julia/blob/d021bc83a1402e30d68cdf777ccdad169f2a9ac9/base/abstractarray.jl#L1291-L1297

**Comment by StefanKarpinski** [comment]
These functions need clearer names. The pmap function gets a pass because that's a fairly standard distributed computing name.

**Comment by samoconnor** [comment]
> These functions need clearer names. The pmap function gets a pass because that's a fairly standard distributed computing name.

OK (the original names were simply an attempt to follow the `pmap` pattern).

@bjarthur [suggested in #14843](https://github.com/JuliaLang/julia/issues/14843#issuecomment-183683746):
> ... get rid of amap and amap! and rename the proposed imap to amap. to get a collection, the user would have to manually say collect(amap), just like they have to do for a dict's keys() and values().

This seems sensible to me (shorthand for `collect(asyncmap())` could still be added later if that was seen as useful).

This PR now adds only one function:

**asyncmap(f, c...; ntasks=100) -> iterator**

Apply f to each element of c using at most 100 asynchronous tasks. For multiple collection arguments, apply f elementwise. Note: `collect(asyncmap(f, c...; ntasks=1))` is equivalent to `map(f, c...)`.

**Comment by samoconnor** [comment]
bump?

**Comment by amitmurthy** [comment]
Looks good. While it is not exported, can we have a different name for `asyncmap` since it is returning an iterator and not a mapped collection? Especially with both `map` and `pmap` returning mapped collections.

File: base/mapiterator.jl [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+#-------------------------------------------------------------------------------
+# MapIterator
+#
+# Simple synchronous map iterator.
+#
+#   collect(MapIterator(f, c...)) == map(f, c...)
+#
+#-------------------------------------------------------------------------------
+
+import Base: start, done, next
```
Comment (amitmurthy) [inline base/mapiterator.jl]: Not required?

**Comment by samoconnor** [comment]
> ... a different name for asyncmap since it is returning an iterator and not a mapped collection?

What about just calling it `StreamMapIterator` instead of having `asyncmap(f, c...; ntasks=0) = StreamMapIterator(f, c...; ntasks=ntasks)` ?

_(I think `map()` should return an iterator to be consistent with `zip()`, `keys()`, `filter()` etc, but that is a seperate discussion.)_

**Comment by amitmurthy** [comment]
That sounds right. BTW, `filter` does return a collection, but I too was initially thrown off by `zip` and `keys` returning iterators.

**Comment by samoconnor** [comment]
`filter` returns an iterator if the input is an iterator...

```julia
julia> typeof(filter(i->true, keys(Dict(1=>1))))
Filter{Function,Base.KeyIterator{Dict{Int64,Int64}}}

julia> typeof(filter(i->true, [1,2,3]))
Array{Int64,1}
```

_(to me that seems really weird because you sometimes need to do `collect(filter(f, c))` and sometimes not so you have to check the tarp of `c` before calling `filter`)_

**Comment by bjarthur** [comment]
functions are supposed to be lowercase according to the style conventions:

http://docs.julialang.org/en/release-0.4/manual/style-guide/#use-naming-conventions-consistent-with-julia-s-base

also, why no exports?

at some point should be documented here:

http://docs.julialang.org/en/release-0.4/manual/parallel-computing/
http://docs.julialang.org/en/release-0.4/stdlib/parallel/?highlight=pmap#Base.pmap

**Comment by samoconnor** [comment]
@bjarthur:
> functions are supposed to be lowercase according to the style conventions:

`StreamMapIterator` is a type (_["type names use capitalization and camel case"]( http://docs.julialang.org/en/release-0.4/manual/style-guide/#use-naming-conventions-consistent-with-julia-s-base)_ ).

> also, why no exports?

because https://github.com/JuliaLang/julia/issues/14843#issuecomment-183875554

**Comment by amitmurthy** [comment]
@StefanKarpinski, your thoughts?

**Comment by amitmurthy** [comment]
I'll merge this in a day if there are no objections. It does not add any exports and will enable enhancing `pmap` to be used in place of code similar to 

```
@sync for x in y
  @async f(x)
end
```

File: base/mapiterator.jl [hunk 6]
```diff
@@ -0,0 +1,175 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+
+"""
+    MapIterator(f, c...) -> iterator
+
+`collect(MapIterator(f, c...))` is equivalent to `map(f, c...)`
+"""
+
```
Comment (tkelman) [inline base/mapiterator.jl]: no newlines after docstrings

File: base/mapiterator.jl [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+function done(itr::AsyncMapIterator, state::AsyncMapState)
+    if !state.done && done(itr.arg_enum, state.enum_state)
+        state.done = true
+        Base.sync_end()
+    end
+    return state.done
+end
+
+function next(itr::AsyncMapIterator, state::AsyncMapState)
+
```
Comment (tkelman) [inline base/mapiterator.jl]: don't start function definitions with empty lines

Comment (samoconnor) [inline base/mapiterator.jl]: Really? Why?
The formatting guidelines say "use whitespace to make the code more readable".

File: base/mapiterator.jl (hunk 6 above)
Comment (samoconnor) [inline base/mapiterator.jl]: Is this a hard requirement to make the docstrings machine-readable? or is it an issue of style?

File: base/mapiterator.jl (hunk 7 above)
Comment (tkelman) [inline base/mapiterator.jl]: IMO this is less readable, not serving any purpose and the majority of the rest of the code in base doesn't do this. That sentence is mostly referring to spaces after commas.

File: base/mapiterator.jl (hunk 6 above)
Comment (tkelman) [inline base/mapiterator.jl]: Looks like it's not required to get docstrings to be associated with the following function, but stylistically it makes the association less obvious.

Comment (MichaelHatherly) [inline base/mapiterator.jl]: Yes, any amount of whitespace, as well as commented lines, will get ignored. But as Tony mentions, stylistically it isn't quite as obvious. Maybe worth a mention in #15136.

Also missing a fullstop as the end of the line.

**Comment by bjarthur** [comment]
@amitmurthy when/who will enhance `pmap` thusly?  i still feel that the full change to the `pmap` API should happen in one PR.

**Comment by amitmurthy** [comment]
Both AV tests timed out. Known issue? How do I restart AV builds?

**Comment by samoconnor** [comment]
@amitmurthy looks like AV is happy now.

**Comment by JeffBezanson** [comment]
The `MapIterator` type is a near-exact duplicate of the `Generator` type that was already added. I would be ok with renaming Generator to `MapIterator` if folks prefer that name.

**Comment by tkelman** [comment]
The `Generator` name is better. We could rename the others here `AsyncGenerator` and `StreamGenerator` if it's really the exact same underlying thing.

**Comment by amitmurthy** [comment]
Yes, the `Generator` name is definitely better. @samoconnor will you take a shot at merging the implementations? Else, I can do it.

**Comment by samoconnor** [comment]
As far as I can see the only difference is that `MapIterator` follows the `map` pattern of taking multiple collections `c...` and using `zip(c...)` to iterate over all of them in parallel.

If this was added to `Generator` you'd have:
```julia
immutable Generator{I,F}
    f::F
    iter::I
end

Generator(f, c...) = Generator(f, zip(c...))

start(g::Generator) = start(g.iter)
done(g::Generator, s) = done(g.iter, s)
function next(g::Generator, s)
    v, s2 = next(g.iter, s)
    g.f(v...), s2
end
```

**Comment by amitmurthy** [comment]
Yes, that is the right thing do do. And if we are going with the `Generator` name, also move `AsyncGenerator` and `StreamGenerator` implementations to `generator.jl`

**Comment by JeffBezanson** [comment]
I find it a bit odd to make the iterator wrapped by Generator always a Zip. It makes it harder to examine what the generator is iterating over.

**Comment by amitmurthy** [comment]
`Generator(f, c1, c...) = Generator(f, zip(c1, c...))` then? Only for mapping functions that take multiple arguments. Like `pmap`.

**Comment by JeffBezanson** [comment]
Yes, I think

```
Generator(f, c1, c...) = Generator(a->f(a...), zip(c1, c...))
```

is probably ok.

**Comment by samoconnor** [comment]
The thing is that `g.f(v...), s2` depends on `v` being a `Tuple`.
So, if you're going to special-case single collection, you need a special-case `next` as well.

**Comment by JeffBezanson** [comment]
With the definition I wrote above `g.f(v)` in `next` is sufficient.

**Comment by samoconnor** [comment]
Ah, yes, our messages crossed somewhere under the sea near Honolulu.

**Comment by JeffBezanson** [comment]
Light is so slow.

**Comment by samoconnor** [comment]
Should I roll a change along these lines into #15409? Or will someone with commit access just do it without a PR?

**Comment by JeffBezanson** [comment]
I'll change it.

**Pull request merged**